File: OriginNG/OriginNG/NGClasses/ANGraph.m
//
//  ANGraph.m
//  AnnieVisualyticsPageView
//
//  Created by hui wang on 1/17/16.
//  Copyright (c) 2016 AnnieLyticx. All rights reserved.
//

#import "ANGraph.h"

#import "AImage_LifeStyle0.h"
// factory for hot spot node path
#import "ANSymbolFactory.h"


@implementation ANGraph
{
    @private
    
    // root of embeded node tree.
    // caller can directly attach a ANLayoutNode tree by passing a root.  an empty ANGraphNodeLayer object will be crated for each node for reserving the drawing layer
    ANLayoutNode* _root;
    
    // array of ANGraphNodeLayer stored in _root
    NSMutableArray* _allNodes;
}

-(id)init
{
    self = [super init];
    
    if (self == nil)
    {
        return nil;
    }
    
    _allNodes = [[NSMutableArray alloc] init];
    
    return self;
}

// access the root object
-(ANLayoutNode*)getRoot
{
    return _root;
}

// functions for crating and populating graph tree of ANGraphNodeLayer node
// starting a new graph
-(ANLayoutNode*)createNewGraph:(APageLayerPath*) layer
{
    _root = [[ANLayoutNode alloc] init];
    _allNodes = [[NSMutableArray alloc] init];
    
    // create {node, layer} object and add onto _allNodes
    _root.index = _allNodes.count;
    [_allNodes addObject:[[ANGraphNodeLayer alloc] initWith:_root layer:layer]];
    
    return _root;
}

// add node
-(ANLayoutNode*)addNode:(ANLayoutNode*) parent index:(int) index weight:(float)weight layer:(APageLayerPath*) layer
{
    if (parent == nil)
    {
        return nil;
    }
    
    // create a new node
    ANLayoutNode* node = [parent addNode:index weight:weight];
    
    if (_allNodes == nil)
    {
        _allNodes = [[NSMutableArray alloc] init];
    }
    
    // create {node, layer} object and add onto _allNodes
    node.index = _allNodes.count;
    [_allNodes addObject:[[ANGraphNodeLayer alloc] initWith:node layer:layer]];
    
    return node;
}

// remove node
-(void)removeNode:(ANLayoutNode*) parent node:(ANLayoutNode*)node
{
    if (parent == nil || node == nil)
    {
        return ;
    }
    
    // remove from _allNodes
    if (_allNodes != nil && node.index >= 0 && node.index < _allNodes.count)
    {
        _allNodes[node.index] = nil;
    }
    [parent removeNode:node];
}

// access to nodes synced inside array
-(int)numberOfNodes
{
    return _allNodes.count;
}

-(ANGraphNodeLayer*)node:(int)index
{
    if (index >= _allNodes.count)
    {
        return nil;
    }
    
    return [_allNodes objectAtIndex:index];
}

-(void)_syncUp:(ANLayoutNode*)node
     nodeArray:(NSMutableArray*)allNodes
{
    if (node == nil || allNodes == nil)
    {
        return ;
    }
    
    node.index = allNodes.count;
    [allNodes addObject:[[ANGraphNodeLayer alloc] initWith:node]];
    
    for (ANLayoutNode* obj in node.childern)
    {
        [self _syncUp:obj nodeArray:allNodes];
    }
}

// sync up node index.  !!! an internal dynamic array of ANGraphNodeLayer is maintainted internally.  the array is for the purpose of fast accessing to nodes of embeded tree.  as result the index needs to be synced up each time when the tree structure changes
-(void)syncUpNodesIndex
{
    if (_root == nil)
    {
        return ;
    }
    
    if (_allNodes == nil)
    {
        _allNodes = [[NSMutableArray alloc] init];
    }
    
    [self _syncUp:_root nodeArray:_allNodes];
}

// pick a sub-tree
-(ANLayoutNode*)pickSubTree
{
    ANLayoutNode* node = [_root.childern objectAtIndex:(_root.childern.count - 1)];
    
    node = [node.childern objectAtIndex:(node.childern.count - 1)];
    
    return node;
}

// pick a segment
-(ANLayoutNode*)pickSegment
{
    return [_root.childern objectAtIndex:(_root.childern.count - 1)];
}

/////////////////////////////////////////////////////////////////////////////
// functions for test purpose
/////////////////////////////////////////////////////////////////////////////
// create tree of graph with presentation layer
-(void) createTestGraph
{
    ANLayoutNode* root = [self createNewGraph:getRandomSymbolPath_Face()];
    
    int nNodes = 0;
    float weight = 1.0;
    int segment = 0;
    
    // segment Prenatal
    ANLayoutNode* Prenatal = [self addNode:root index:nNodes weight:weight layer:getRandomSymbolPath_Face()];
    Prenatal.segment = segment;
    
    ANLayoutNode* node = [self addNode:Prenatal index:nNodes weight:weight layer:getRandomSymbolPath()];     node.segment = segment;
    
    ANLayoutNode* risks = [self addNode:Prenatal index:nNodes weight:weight layer:getRandomSymbolPath()];
    risks.segment = segment;
    
    node = [self addNode:risks index:nNodes weight:weight layer:getRandomSymbolPath()];
    node.segment = segment;
    node = [self addNode:risks index:nNodes weight:weight layer:getRandomSymbolPath()];
    node.segment = segment;
    
    risks = [self addNode:risks index:nNodes weight:weight layer:getRandomSymbolPath()];
    risks.segment = segment;
    
    node = [self addNode:risks index:nNodes weight:weight layer:getRandomSymbolPath()];
    node.segment = segment;
    node = [self addNode:risks index:nNodes weight:weight layer:getRandomSymbolPath()];
    node.segment = segment;
    
    risks = [self addNode:Prenatal index:nNodes weight:weight layer:getRandomSymbolPath_Face()];
    risks.segment = segment;
    
    node = [self addNode:risks index:nNodes weight:weight layer:getRandomSymbolPath()];
    node.segment = segment;
    node = [self addNode:risks index:nNodes weight:weight layer:getRandomSymbolPath()];
    node.segment = segment;
    
    // segment Infant
    segment += 1;
    ANLayoutNode* Infant = [self addNode:root index:nNodes weight:weight layer:getRandomSymbolPath_Face()];
    [self addNode:Infant index:nNodes weight:weight layer:getRandomSymbolPath_Face()];
    
    risks = [self addNode:Infant index:nNodes weight:weight layer:getRandomSymbolPath_Face()];
    [self addNode:risks index:nNodes weight:weight layer:getRandomSymbolPath()];
    [self addNode:risks index:nNodes weight:weight layer:getRandomSymbolPath()];
    [self addNode:risks index:nNodes weight:weight layer:getRandomSymbolPath()];
    [self addNode:risks index:nNodes weight:weight layer:getRandomSymbolPath()];
    [self addNode:risks index:nNodes weight:weight layer:getRandomSymbolPath()];
    
    // segment Child
    ANLayoutNode* Child = [self addNode:root index:nNodes weight:weight layer:getRandomSymbolPath_Face()];
    [self addNode:Child index:nNodes weight:weight layer:getRandomSymbolPath_Face()];
    
    risks = [self addNode:Child index:nNodes weight:weight layer:getRandomSymbolPath_Face()];
    [self addNode:risks index:nNodes weight:weight layer:getRandomSymbolPath()];
    [self addNode:risks index:nNodes weight:weight layer:getRandomSymbolPath()];
    [self addNode:risks index:nNodes weight:weight layer:getRandomSymbolPath()];
    [self addNode:risks index:nNodes weight:weight layer:getRandomSymbolPath()];
    [self addNode:risks index:nNodes weight:weight layer:getRandomSymbolPath()];
    [self addNode:risks index:nNodes weight:weight layer:getRandomSymbolPath()];
    
    
    // segment Adult
    ANLayoutNode* Adult = [self addNode:root index:nNodes weight:weight layer:getRandomSymbolPath_Face()];
    [self addNode:Adult index:nNodes weight:weight layer:getRandomSymbolPath_Face()];
    
    risks = [self addNode:Adult index:nNodes weight:weight layer:getRandomSymbolPath_Face()];
    [self addNode:risks index:nNodes weight:weight layer:getRandomSymbolPath()];
    [self addNode:risks index:nNodes weight:weight layer:getRandomSymbolPath()];
    [self addNode:risks index:nNodes weight:weight layer:getRandomSymbolPath()];
    [self addNode:risks index:nNodes weight:weight layer:getRandomSymbolPath()];
    [self addNode:risks index:nNodes weight:weight layer:getRandomSymbolPath()];
    [self addNode:risks index:nNodes weight:weight layer:getRandomSymbolPath()];
    [self addNode:risks index:nNodes weight:weight layer:getRandomSymbolPath()];
    [self addNode:risks index:nNodes weight:weight layer:getRandomSymbolPath()];
    [self addNode:risks index:nNodes weight:weight layer:getRandomSymbolPath()];
    [self addNode:risks index:nNodes weight:weight layer:getRandomSymbolPath()];
    [self addNode:risks index:nNodes weight:weight layer:getRandomSymbolPath()];
    [self addNode:risks index:nNodes weight:weight layer:getRandomSymbolPath()];
    
    risks = [self addNode:risks index:nNodes weight:weight layer:getRandomSymbolPath_Face()];
    [self addNode:risks index:nNodes weight:weight layer:getRandomSymbolPath()];
    
    // segment Elderly
    ANLayoutNode* Elderly = [self addNode:root index:nNodes weight:weight layer:getRandomSymbolPath_Face()];
    [self addNode:Elderly index:nNodes weight:weight layer:getRandomSymbolPath_Face()];
    
    risks = [self addNode:Elderly index:nNodes weight:weight layer:getRandomSymbolPath_Face()];
    [self addNode:risks index:nNodes weight:weight layer:getRandomSymbolPath()];
    [self addNode:risks index:nNodes weight:weight layer:getRandomSymbolPath()];
    [self addNode:risks index:nNodes weight:weight layer:getRandomSymbolPath()];
    [self addNode:risks index:nNodes weight:weight layer:getRandomSymbolPath()];
    [self addNode:risks index:nNodes weight:weight layer:getRandomSymbolPath()];
    
    risks = [self addNode:risks index:nNodes weight:weight layer:getRandomSymbolPath_Face()];
    [self addNode:risks index:nNodes weight:weight layer:getRandomSymbolPath()];
    [self addNode:risks index:nNodes weight:weight layer:getRandomSymbolPath()];
    [self addNode:risks index:nNodes weight:weight layer:getRandomSymbolPath()];
}

@end

File: BackUpForABook/UIControls/NGClasses_swift/ANGraph.swift
//
//  ANGraph.swift
//  NGClasses
//
//  Created by iMac on 16/12/13.
//  Copyright © 2016年 LyTsai. All rights reserved.
//

import Foundation
import UIKit

// wrap a ANLayoutNode tree
class ANGraph {
    // root of embeded node tree.
    // caller can directly attach a ANLayoutNode tree by passing a root.  an empty ANGraphNodeLayer object will be crated for each node for reserving the drawing layer
    fileprivate var root: ANLayoutNode!
    
    // array of ANGraphNodeLayer stored in _root
    fileprivate var allNodes = [ANGraphNodeLayer?]()
}

extension ANGraph {
    // access the root object
    func getRoot() -> ANLayoutNode {
        return root
    }
    
    /////////////////////////////////////////////////////////////////////////////
    // functions for creating and populating graph tree of ANGraphNodeLayer node
    // starting a new graph
    func createNewGraph(_ layer: ANNodeLayer) -> ANLayoutNode {
        root = ANLayoutNode()
        
        // create {node, layer} object and add onto _allNodes
        root.index = 0
        allNodes.append(ANGraphNodeLayer(root, layer: layer))
        
        return root
    }
    
    // add node
    func addNode(_ parent: ANLayoutNode!, index: Int, weight: CGFloat, layer: CALayer) -> ANLayoutNode? {
        if parent == nil {
            return nil
        }
        
        // create a new node
        let node = parent.addNode(index, weight: weight)
        
        // create {node, layer} object and add onto _allNodes
        node.index = allNodes.count
        allNodes.append(ANGraphNodeLayer(node, layer: layer))
        
        return node
    }
    
    // remove node
    func removeNode(_ parent: ANLayoutNode!, node: ANLayoutNode!) {
        if parent == nil || node == nil {
            return
        }
        
        // remove from _allNodes
        if allNodes.count != 0 && node.index >= 0 && node.index < allNodes.count {
            allNodes[node.index] = nil
        }
        parent.removeNode(node)
    }
    
    /////////////////////////////////////////////////////////////////////////////
    // access to nodes synced inside array
    func numberOfNodes() -> Int {
        return allNodes.count
    }
    
    func node(_ index: Int) -> ANGraphNodeLayer? {
        if index < 0 || index >= allNodes.count {
            return nil
        }
        
        return allNodes[index]
    }
    
    // call this function to resync after delting many nodes
    fileprivate func syncUp(_ node: ANLayoutNode!, nodeArray allNodes: inout [ANGraphNodeLayer?]) {
        if node == nil {
            return
        }
        
        node.index = allNodes.count
        allNodes.append(ANGraphNodeLayer(node))
        
        for child in node.children {
            syncUp(child, nodeArray: &allNodes)
        }
    }
    
    // sync up node index.  !!! an internal dynamic array of ANGraphNodeLayer is maintainted internally.  the array is for the purpose of fast accessing to nodes of embeded tree.  as result the index needs to be synced up each time when the tree structure changes
    func syncUpNodesIndex() {
        if root == nil { return }
        syncUp(root, nodeArray: &allNodes)
    }
    
    // pick a sub-tree
    func pickSubTree() -> ANLayoutNode? {
        var node = root.children.last
        node = node?.children.last

        return node
    }
 
    // pick a segment
    func pickSegment() -> ANLayoutNode? {
        return root.children.last
    }
    
    /////////////////////////////////////////////////////////////////////////////
    // functions for test purpose
    /////////////////////////////////////////////////////////////////////////////
    // create tree of graph with presentation layer
    func createTestGraph() {
        let root = createNewGraph(getRandomSymbolPath_Face())
        
        let nNodes: Int = 0
        let weight: CGFloat = 1
        var segment: Int = 0
        
        // segment Prenatal
        let prenatal = addNode(root, index: nNodes, weight: weight, layer: getRandomSymbolPath_Face())!
        prenatal.segment = segment
 
        var node = addNode(prenatal, index: nNodes, weight: weight, layer: getRandomSymbolPath())!
        node.segment = segment
        var risks = addNode(prenatal, index: nNodes, weight: weight, layer: getRandomSymbolPath())!
        risks.segment = segment

        node = addNode(risks, index: nNodes, weight: weight, layer: getRandomSymbolPath())!
        node.segment = segment
        node = addNode(risks, index: nNodes, weight: weight, layer: getRandomSymbolPath())!
        node.segment = segment
        
        risks = addNode(risks, index: nNodes, weight: weight, layer: getRandomSymbolPath())!
        risks.segment = segment
        
        node = addNode(risks, index: nNodes, weight: weight, layer: getRandomSymbolPath())!
        node.segment = segment
        node = addNode(risks, index: nNodes, weight: weight, layer: getRandomSymbolPath())!
        node.segment = segment
        
        risks = addNode(prenatal, index: nNodes, weight: weight, layer: getRandomSymbolPath_Face())!
        risks.segment = segment
        
        node = addNode(risks, index: nNodes, weight: weight, layer: getRandomSymbolPath_Face())!
        node.segment = segment
        node = addNode(risks, index: nNodes, weight: weight, layer: getRandomSymbolPath_Face())!
        node.segment = segment
      
        // segment Infant
        segment += 1
        let infant = addNode(root, index: nNodes, weight: weight, layer: getRandomSymbolPath_Face())!
        let _ = addNode(infant, index: nNodes, weight: weight, layer: getRandomSymbolPath_Face())
        
        risks = addNode(infant, index: nNodes, weight: weight, layer: getRandomSymbolPath_Face())!
        let _ = addNode(risks, index: nNodes, weight: weight, layer: getRandomSymbolPath())
        let _ = addNode(risks, index: nNodes, weight: weight, layer: getRandomSymbolPath())
        let _ = addNode(risks, index: nNodes, weight: weight, layer: getRandomSymbolPath())
        let _ = addNode(risks, index: nNodes, weight: weight, layer: getRandomSymbolPath())
        let _ = addNode(risks, index: nNodes, weight: weight, layer: getRandomSymbolPath())
   
        // segment Child
        let child = addNode(root, index: nNodes, weight: weight, layer: getRandomSymbolPath_Face())!
        let _ = addNode(child, index: nNodes, weight: weight, layer: getRandomSymbolPath_Face())
        
        risks = addNode(child, index: nNodes, weight: weight, layer: getRandomSymbolPath_Face())!
        let _ = addNode(risks, index: nNodes, weight: weight, layer: getRandomSymbolPath())
        let _ = addNode(risks, index: nNodes, weight: weight, layer: getRandomSymbolPath())
        let _ = addNode(risks, index: nNodes, weight: weight, layer: getRandomSymbolPath())
        let _ = addNode(risks, index: nNodes, weight: weight, layer: getRandomSymbolPath())
        let _ = addNode(risks, index: nNodes, weight: weight, layer: getRandomSymbolPath())
        let _ = addNode(risks, index: nNodes, weight: weight, layer: getRandomSymbolPath())
        
        // segment Adult
        let adult = addNode(root, index: nNodes, weight: weight, layer: getRandomSymbolPath_Face())!
        let _ = addNode(adult, index: nNodes, weight: weight, layer: getRandomSymbolPath_Face())
        
        risks = addNode(adult, index: nNodes, weight: weight, layer: getRandomSymbolPath_Face())!
        let _ = addNode(risks, index: nNodes, weight: weight, layer: getRandomSymbolPath())
        let _ = addNode(risks, index: nNodes, weight: weight, layer: getRandomSymbolPath())
        let _ = addNode(risks, index: nNodes, weight: weight, layer: getRandomSymbolPath())
        let _ = addNode(risks, index: nNodes, weight: weight, layer: getRandomSymbolPath())
        let _ = addNode(risks, index: nNodes, weight: weight, layer: getRandomSymbolPath())
        let _ = addNode(risks, index: nNodes, weight: weight, layer: getRandomSymbolPath())
        let _ = addNode(risks, index: nNodes, weight: weight, layer: getRandomSymbolPath())
        let _ = addNode(risks, index: nNodes, weight: weight, layer: getRandomSymbolPath())
        let _ = addNode(risks, index: nNodes, weight: weight, layer: getRandomSymbolPath())
        let _ = addNode(risks, index: nNodes, weight: weight, layer: getRandomSymbolPath())
        let _ = addNode(risks, index: nNodes, weight: weight, layer: getRandomSymbolPath())
        let _ = addNode(risks, index: nNodes, weight: weight, layer: getRandomSymbolPath())

        risks = addNode(risks, index: nNodes, weight: weight, layer: getRandomSymbolPath_Face())!
        let _ = addNode(risks, index: nNodes, weight: weight, layer: getRandomSymbolPath())
   
        // segment Elderly
        let elderly = addNode(root, index: nNodes, weight: weight, layer: getRandomSymbolPath_Face())!
        let _ = addNode(elderly, index: nNodes, weight: weight, layer: getRandomSymbolPath_Face())
        
        risks = addNode(elderly, index: nNodes, weight: weight, layer: getRandomSymbolPath_Face())!
        let _ = addNode(risks, index: nNodes, weight: weight, layer: getRandomSymbolPath())
        let _ = addNode(risks, index: nNodes, weight: weight, layer: getRandomSymbolPath())
        let _ = addNode(risks, index: nNodes, weight: weight, layer: getRandomSymbolPath())
        let _ = addNode(risks, index: nNodes, weight: weight, layer: getRandomSymbolPath())
        let _ = addNode(risks, index: nNodes, weight: weight, layer: getRandomSymbolPath())
        
        risks = addNode(risks, index: nNodes, weight: weight, layer: getRandomSymbolPath_Face())!
        let _ = addNode(risks, index: nNodes, weight: weight, layer: getRandomSymbolPath())
        let _ = addNode(risks, index: nNodes, weight: weight, layer: getRandomSymbolPath())
        let _ = addNode(risks, index: nNodes, weight: weight, layer: getRandomSymbolPath())

    }
}
